turn/src/server/mod.rs

```diff
@@ -52,40 +56,80 @@ impl Server {
             let auth_handler = Arc::clone(&s.auth_handler);
             let realm = s.realm.clone();
             let channel_bind_timeout = s.channel_bind_timeout;
-            let shutdown_rx = shutdown_rx.clone();
-
-            tokio::spawn(async move {
-                let allocation_manager = Arc::new(Manager::new(ManagerConfig {
-                    relay_addr_generator: p.relay_addr_generator,
-                }));
-
-                Server::read_loop(
-                    p.conn,
-                    allocation_manager,
-                    nonces,
-                    auth_handler,
-                    realm,
-                    channel_bind_timeout,
-                    shutdown_rx,
-                )
-                .await;
-            });
+            let handle_rx = handle.subscribe();
+            let conn = p.conn;
+            let allocation_manager = Arc::new(Manager::new(ManagerConfig {
+                relay_addr_generator: p.relay_addr_generator,
+            }));
+
+            tokio::spawn(Server::read_loop(
+                conn,
+                allocation_manager,
+                nonces,
+                auth_handler,
+                realm,
+                channel_bind_timeout,
+                handle_rx,
+            ));
         }
 
         Ok(s)
     }
 
+    /// Deletes all existing [`crate::allocation::Allocation`]s by the provided `username`.
+    pub async fn delete_allocations_by_username(&self, username: String) -> Result<()> {
+        let tx = self.handle.lock().await.clone();
+        if let Some(tx) = tx {
+            let (closed_tx, closed_rx) = mpsc::channel(1);
+            tx.send(Command::DeleteAllocations(username, Arc::new(closed_rx)))
+                .map_err(|_| Error::ErrClosed)?;
+
+            closed_tx.closed().await;
+
+            Ok(())
+        } else {
+            Err(Error::ErrClosed)
+        }
+    }
+
     async fn read_loop(
         conn: Arc<dyn Conn + Send + Sync>,
         allocation_manager: Arc<Manager>,
         nonces: Arc<Mutex<HashMap<String, Instant>>>,
         auth_handler: Arc<dyn AuthHandler + Send + Sync>,
         realm: String,
         channel_bind_timeout: Duration,
-        mut shutdown_rx: watch::Receiver<bool>,
+        mut handle_rx: broadcast::Receiver<Command>,
     ) {
         let mut buf = vec![0u8; INBOUND_MTU];
 
+        let (close_tx, mut close_rx) = mpsc::channel(1);
```

Comment:
Any reason these channels to communicate a single thing aren't `oneshot`? Also, it would be more suitable to use `oneshot::channel:<T>` where `T` is a return value, even if it's just `()`, and then use the sender being deallocated to detect shutdown